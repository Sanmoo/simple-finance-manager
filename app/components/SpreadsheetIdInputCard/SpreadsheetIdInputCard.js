import React, { useState } from 'react';
import PropTypes from 'prop-types';
import CircularProgress from '@material-ui/core/CircularProgress';
import styled from 'styled-components';
import { withTheme } from '@material-ui/core/styles';
import {
  Typography,
  Button,
  Card,
  CardContent,
  CardActions,
  TextField as TF,
} from '@material-ui/core';

const TextField = styled(TF)`
  width: 100%;
`;

const LoadingIcon = styled(CircularProgress)`
  margin: ${p => p.theme.spacing.unit};
`;

function SpreadsheetIdInputCard({
  onSubmitSId,
  onCreateNewSpreadsheet,
  isLoading,
  theme,
}) {
  const [sId, setId] = useState('');

  return (
    <Card>
      <CardContent>
        <Typography variant="h4" component="h2">
          Please provide your spreadsheet Id
        </Typography>
        <TextField
          id="spreadsheetId"
          label="Your spreadsheet Id"
          value={sId}
          onChange={evt => setId(evt.target.value)}
          margin="normal"
        />
        <Typography variant="subheading" component="p">
          {"Or click 'Create one for me' if you don't have one yet"}
        </Typography>
      </CardContent>
      <CardActions>
        {isLoading && <LoadingIcon theme={theme} size={20} />}
        <Button
          size="small"
          onClick={() => onSubmitSId(sId)}
          disabled={isLoading || !sId}
        >
          OK
        </Button>
        <Button
          size="small"
          onClick={() => onCreateNewSpreadsheet()}
          disabled={isLoading}
        >
          Create one for me
        </Button>
      </CardActions>
    </Card>
  );
}

SpreadsheetIdInputCard.propTypes = {
  onSubmitSId: PropTypes.func.isRequired,
  onCreateNewSpreadsheet: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  theme: PropTypes.object.isRequired,
};

export default withTheme()(SpreadsheetIdInputCard);
