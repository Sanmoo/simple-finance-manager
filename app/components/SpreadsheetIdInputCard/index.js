import SpreadsheetIdInputCard from './SpreadsheetIdInputCard';

export default SpreadsheetIdInputCard;
