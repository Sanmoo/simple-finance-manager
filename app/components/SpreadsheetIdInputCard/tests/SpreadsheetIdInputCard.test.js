/**
 *
 * Tests for SpreadsheetIdInputCard
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import 'jest-dom/extend-expect';
import { SpreadsheetIdInputCard } from '../SpreadsheetIdInputCard';

describe('<SpreadsheetIdInputCard />', () => {
  let props;
  let firstChild;
  let getByText;
  let getByLabelText;
  const renderComponent = () => {
    ({
      container: { firstChild },
      getByText,
      getByLabelText,
    } = render(<SpreadsheetIdInputCard {...props} />));
  };

  beforeEach(() => {
    props = {
      onSubmitSId: jest.fn(),
      onCreateNewSpreadsheet: jest.fn(),
      isLoading: false,
      theme: {
        spacing: { unit: 1 },
      },
    };
  });

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    renderComponent();
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    renderComponent();
    expect(firstChild).toMatchSnapshot();
  });

  it('should show button OK disabled', () => {
    renderComponent();
    expect(getByText('OK')).toBeDisabled();
  });

  describe('when I type a spreadsheet Id', () => {
    beforeEach(() => {
      renderComponent();
      fireEvent.change(getByLabelText('Your spreadsheet Id'), {
        target: { value: 'something' },
      });
    });

    it('should show button OK enabled', () => {
      expect(getByText('OK')).not.toBeDisabled();
    });

    describe('when ok is clicked', () => {
      beforeEach(() => {
        fireEvent.click(getByText('OK'));
      });

      it('submit callback is called', () => {
        expect(props.onSubmitSId).toHaveBeenCalledWith('something');
      });
    });
  });

  describe('when I click on the create one for me button', () => {
    beforeEach(() => {
      renderComponent();
      fireEvent.click(getByText('Create one for me'));
    });

    it('calls prop onCreateNewSpreadsheet back', () => {
      expect(props.onCreateNewSpreadsheet).toHaveBeenCalledTimes(1);
    });
  });
});
