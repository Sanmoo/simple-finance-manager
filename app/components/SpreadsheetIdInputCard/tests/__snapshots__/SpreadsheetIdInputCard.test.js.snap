// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpreadsheetIdInputCard /> Should render and match the snapshot 1`] = `
<div
  class="MuiPaper-root-158 MuiPaper-elevation1-161 MuiPaper-rounded-159 MuiCard-root-157"
>
  <div
    class="MuiCardContent-root-185"
  >
    <h2
      class="MuiTypography-root-186 MuiTypography-h4-201"
    >
      Please provide your spreadsheet Id
    </h2>
    <div
      class="MuiFormControl-root-222 MuiFormControl-marginNormal-223 SpreadsheetIdInputCard__TextField-ocjwh8-0 ihlGeT"
    >
      <label
        class="MuiFormLabel-root-237 MuiInputLabel-root-226 MuiInputLabel-formControl-231 MuiInputLabel-animated-234"
        data-shrink="false"
        for="spreadsheetId"
      >
        Your spreadsheet Id
      </label>
      <div
        class="MuiInputBase-root-257 MuiInput-root-244 MuiInput-underline-248 MuiInputBase-formControl-258 MuiInput-formControl-245"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input-267 MuiInput-input-252"
          id="spreadsheetId"
          type="text"
          value=""
        />
      </div>
    </div>
    <p
      class="MuiTypography-root-186 MuiTypography-subheading-193"
    >
      Or click 'Create one for me' if you don't have one yet
    </p>
  </div>
  <div
    class="MuiCardActions-root-274"
  >
    <button
      class="MuiButtonBase-root-303 MuiButtonBase-disabled-304 MuiButton-root-277 MuiButton-text-279 MuiButton-flat-282 MuiButton-sizeSmall-300 MuiButton-disabled-297 MuiCardActions-action-276"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label-278"
      >
        OK
      </span>
    </button>
    <button
      class="MuiButtonBase-root-303 MuiButton-root-277 MuiButton-text-279 MuiButton-flat-282 MuiButton-sizeSmall-300 MuiCardActions-action-276"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label-278"
      >
        Create one for me
      </span>
      <span
        class="MuiTouchRipple-root-306"
      />
    </button>
  </div>
</div>
`;
